DM-25228: Create README for sample alert packets (#1)
lsst-dm/sample_alert_info: README.md

# Rubin Observatory Sample Alerts

The Vera C. Rubin Observatory is releasing sample alert packets produced by its pipelines run on precursor survey data.
These alerts are intended to provide candidate community alert brokers as well as science users the chance to preview the alert format and contents and to provide feedback to the project.

## Obtaining the Data

Sample alert data may be obtained at [this link](https://lsst.ncsa.illinois.edu/~ebellm/sample_precursor_alerts/).
We anticipate providing reprocessed and updated alert samples as pipeline development progresses, so the data are arranged in a directory structure:

```
    sample_precursor_alerts/
	YYYY-MM-DD/
	    dataset/
		all_visits_{dataset}_{date}.avro
		association_{dataset}_{date}.db
		single_ccd_sample_{dataset}_{date}.avro
		single_visit_sample_{dataset}_{date}.avro
```

### Alert packets

Alerts are provided as Snappy-compressed [Avro](https://avro.apache.org/) files.
The schema used to store the alerts is included in the file.

> As the complete sample is about 1 TB, We include two smaller subsets--alerts from a single CCD-visit, and all of the alerts from all CCDs in a single visit.

Reviewer: ```suggestion
As the complete sample is about 1 TB, we include two smaller subsets -- alerts from a single CCD-visit, and all of the alerts from all CCDs in a single visit.
```

> ### Database

Reviewer 2: (Missing blank line)

> In addition to accessing alert packets via community brokers, during the LSST survey science users with data rights will be able to query the relational Prompt Products Database (PPDB).  

Reviewer 2: ```suggestion
In addition to accessing alert packets via community brokers and through the project-provided Alert Filtering Service, during the survey members of the data rights community will be able to query the relational Prompt Products Database (PPDB). 
```

> In addition to accessing alert packets via community brokers, during the LSST survey science users with data rights will be able to query the relational Prompt Products Database (PPDB).  

Reviewer 3: It's still Apdb correct? If so you've got more than a few of these to change name wise.

Author: End users will access the PPDB rather than the APDB, so I used it deliberately throughout.

> The contents of the alerts are taken directly from the PPDB `DIASource`, `DIAForcedSource`, `DIAObject`, and `SSObject` records.

Reviewer 2: ```suggestion
The contents of the alerts are correspond directly to the `DIASource`, `DIAForcedSource`, `DIAObject`, and `SSObject` records.
```
(The alerts aren't actually generated from the PPDB)

For some use cases direct queries to the PPDB may be more efficient than processing alert packets.

To allow users to explore these different modes of data access, we are also providing the PPDB (`association_{dataset}_{date}.db`) from which the sample alerts were generated as an [SQLite](https://www.sqlite.org) database.
The SQLite database can be queried using SQL through a variety of interfaces.

Note that in operations, users will query the PPDB using [ADQL](http://www.ivoa.net/documents/ADQL/) via a [TAP service](http://www.ivoa.net/documents/TAP/).

> ### Software

Reviewer 2: (Missing blank line)

Reviewer 2: (The document is pretty inconsistent throughout about whether headings are followed by blank lines; not important, but makes me twitch. I won't flag each time...!)

> Users should install [lsst.alert.packet](https://github.com/lsst/alert_packet) to parse alerts and optionally [alert-stream-simulator](https://github.com/lsst-dm/alert-stream-simulator/) and [alert-stream](https://github.com/lsst-dm/alert_stream) if they wish to simulate a Kafka alert stream.

Reviewer 2: I think we need a bit more explanation here. In particular:

- https://github.com/lsst-dm/alert-stream-simulator talks about running `make` to download datasets from `~swnelson`. Is that in addition to the datasets described in this file? How do they fit together?
- Both lsst-dm/alert_stream and lsst-dm/alert-stream-simulator talk about using `docker-compose`, etc, to get up and running. Are users supposed to follow both of those sets of instructions? Does one take precedence?

To be fair, I'm not sure exactly what other documentation you plan to provide to end users. If this file is all that's coming, then I think it needs to address questions like the above. If there's a separate “software reference” that explains how these packages fit together, then the above points are moot.

Reviewer: > https://github.com/lsst-dm/alert-stream-simulator talks about running make to download datasets from ~swnelson. Is that in addition to the datasets described in this file? How do they fit together?

I vote that we should change `alert-stream-simulator`'s Makefile to get the datasets from Eric's directory.

> Both lsst-dm/alert_stream and lsst-dm/alert-stream-simulator talk about using docker-compose, etc, to get up and running. Are users supposed to follow both of those sets of instructions? Does one take precedence?

Almost everything in `lsst-dm/alert_stream`'s README is obsolete and subsumed by `alert-stream-simulator`, I think.

Author: I see this README as focused on the sample alerts themselves; usage instructions for `alert-stream-simulator` etc. should live with those packages.

My suggestion would be:

- I put a symlink at the root of the data directory (`https://lsst.ncsa.illinois.edu/~ebellm/sample_precursor_alerts/`) for the latest single CCD and single visit datasets (and full dataset, I guess), and `alert-stream-simulator` can pull those via the Makefile
- We update the  `alert-stream-simulator` README to describe how to bring in larger alert samples (download them into `data/`, which is simple enough, but explicit instructions are good)
- We wipe out everything in the `alert_stream` README that we're not planning to use or publicly support (at least right now).

## External References and Additional Resources

> The [Data Products Definition Document (DPDD)](http://ls.st/dpdd) provides a change-controlled guide to what scientific contents are planned for alerts and other Rubin Observatory data products.

Reviewer 3: Does the latest DPDD include the DiaForcedSource table?

Author: Yes (Table 2)

[LDM-612](http://ls.st/LDM-612) describes "Plans and Policies for LSST Alert Distribution," including a high-level overview of the relevant components, a summary of data rights issues, and extensive discussion of the role and selection of community alert brokers.

[DMTN-093](https://dmtn-093.lsst.io/) presents the technical design of the alert distribution system.

[DMTN-102](https://dmtn-102.lsst.io/) summarizes "Key Numbers" for alerts, estimating numbers, sizing, etc.

[lsst.alert.packet](https://github.com/lsst/alert_packet) is a Python package that contains alert schemas as well as utilities for reading and writing Rubin alert packets.

> Community broker authors will be interested in simulating the Kafka alert stream itself.  [alert-stream-simulator](https://github.com/lsst-dm/alert-stream-simulator/) provides the relevant code; the overall design is discussed in 

Reviewer 2: Not important, but the line breaking gets a little funky here.

[DMTN-149](https://dmtn-149.lsst.io/).

> [alert-stream](https://github.com/lsst-dm/alert_stream) currently contains additional code for converting from the [Confluent Wire Format](https://dmtn-093.lsst.io/#management-and-evolution) used in the Kafka stream.

Reviewer 2: Underscore vs hypen in the package name.

(It's unfortunate that we have alert-stream-simulator but alert_stream... 🤷‍♂️)

## Data Sources

### DECam-HiTS

Sample alerts are provided from a subset of the DECam High Cadence Transient Search (HiTS; [Förster et al. 2016](https://ui.adsabs.harvard.edu/abs/2016ApJ...832..155F/abstract)).

We process g-band data from three fields (`Blind15A_26`, `Blind15A_40`, and `Blind15A_42`) from the 2015 campaign and use coadded templates built from the matching fields in the 2014 campaign.  
In total there are 5020 processed CCD-visits.

Users interested in performing their own reductions can obtain these data [here](https://github.com/lsst/ap_verify_hits2015).
The `ap_verify` package in the LSST Stack can be used to [process this dataset](https://pipelines.lsst.io/modules/lsst.ap.verify/running.html).

> ## Limitations and Caveats

Reviewer: This section is good and detailed. It would be nice if it could appear alongside the data permanently to, like on `lsst-web` alongside the `.avro` files, or perhaps linked to from a changelog, or something.

If a user goes and gets an old dataset, and then looks on the README and sees the latest notes, they may get confused.

Author: I was planning to include the appropriate version of the README alongside the avro files, as you suggest.

The Alert Production pipelines are not yet complete, so not all of the alert contents prescribed by the DPDD are populated in these sample alerts.

> Missing fields will have `null` entries.

Reviewer 3: Missing fields and data that failed to be fit/processed will have `null` entries.

We describe the most significant omissions below.

### Sample
Each `DIASource` detected at or above 5 sigma in the difference image produces a `DIASource` entry in the PPDB and a corresponding alert packet.
Because detection SNR is different than the measured SNR of a point-source fit, not all `DIASources` will have `psFlux/psFluxErr > 5`.

The sample is not filtered for artifacts or bad subtractions.
Through visual inspection we suggest that the current proportion of artifacts to real transients is about 4:1.
We expect continued improvement in this ratio.

### Formats
The Avro format, schema structure, and alert contents released here reflect the current baseline, but users should be prepared for continued evolution and potentially breaking changes.

> The top-level `AlertID` record is currently not populated.

Reviewer: `alertId` is not a nullable field. What do we put in there?

Some nits on phrasing:

```suggestion
The top-level `alertId` field is currently not populated.
```

Author: Currently it's just the DIASource id, but  DM-24858 will change that.  I'll update the README.

https://github.com/lsst/ap_association/blob/87898ef4749329fb0c47ff2995ee90c5b53f7ae5/python/lsst/ap/association/packageAlerts.py#L147

### `DIASource` Records
* The detection signal-to-noise (`SNR`) is not populated 
* Trailed source fits are not implemented
* No extendedness or spuriousness score is included
	
### `DIAObject` Records

* Parallax and proper motion fitting is not performed

> * Timeseries features are not computed

Reviewer 3: This is not true right? We do have some Static time series features implemented. Just not the LS columns.

Author: We are not propagating them into the alerts themselves, though, correct?  All of the `Lc(Non)Periodic` entries are `null`.

Reviewer 3: Ah, you could write the measurements we currently do into the schema though. Those would produce some of the non-periodic measurements.

Author: I'm inclined to save that for the future--both because I want to get this sample out the door and because my next™ project is to work on the timeseries features further.

Author: I'll just say "Timeseries features are not _included_"
